File: Programação orientada a objetos/POO 241 - Trab 1 - Pedro Quedevez/Docente.java
import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;

public class Docente {
    private Long codigo;
    private String nome;
    private LocalDate dataNascimento;
    private LocalDate dataIngresso;
    private ArrayList<Ocorrencia> ocorrencias;
    private ArrayList<Publicacao> publicacoes;

    public static ArrayList<Long> codigosCadastrados = new ArrayList<Long>();

    public Docente(String csvline) throws excecao {
        String[] campos = csvline.split(";");
        try {
            if (campos.length == 4) {
                if (codigosCadastrados.contains(Long.parseLong(campos[0].trim()))) {
                    throw new excecao("Código repetido para docente: " + campos[0] + ".");

                }
                codigosCadastrados.add(Long.parseLong(campos[0].trim()));
                this.codigo = Long.parseLong(campos[0]);
                this.nome = campos[1].trim();
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
                this.dataNascimento = LocalDate.parse(campos[2].trim(), formatter);
                this.dataIngresso = LocalDate.parse(campos[3].trim(), formatter);
                this.ocorrencias = new ArrayList<Ocorrencia>();
                this.publicacoes = new ArrayList<Publicacao>();
            }
        } catch (java.lang.NumberFormatException e) {
            throw new excecao("Erro de formatacao");
        }
    }

    public int idadeDocente(int anoAtual) {

        LocalDate dataAtual = LocalDate.of(anoAtual, 1, 1);
        Period periodo = Period.between(dataNascimento, dataAtual);
        return periodo.getYears();

    }

    public int tempoIngresso(int anoAtual) {
        LocalDate dataAtual = LocalDate.of(anoAtual, 1, 1);
        Period periodo = Period.between(dataIngresso, dataAtual);
        return periodo.getYears();
    }

    public boolean possuiPubAnosAnteriores(int anoAtual, int qntAnosValidos, String qualiLimite, int qtnMinimaPub) {
        int anoLimite = anoAtual - qntAnosValidos;
        int qtdaux = qtnMinimaPub;
        ArrayList<Publicacao> publicacoes = this.getPublicacoes();
        for (Publicacao publicacao : publicacoes) {
            if (publicacao.getAno() >= anoLimite
                    && ItemListaPublicacoes.comparaQualis(publicacao.getQuali(), qualiLimite)) {
                qtdaux--;
            }
        }
        if (qtdaux <= 0) {
            return true;
        }
        return false;
    }

    @Override
    public String toString() {
        return codigo + ";" + nome + ";" + dataNascimento + ";" + dataIngresso;
    }

    public Long getCodigo() {
        return codigo;
    }

    public void setCodigo(Long codigo) {
        this.codigo = codigo;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public LocalDate getDataNascimento() {
        return dataNascimento;
    }

    public void setDataNascimento(LocalDate dataNascimento) {
        this.dataNascimento = dataNascimento;
    }

    public LocalDate getDataIngresso() {
        return dataIngresso;
    }

    public void setDataIngresso(LocalDate dataIngresso) {
        this.dataIngresso = dataIngresso;
    }

    public ArrayList<Ocorrencia> getOcorrencias() {
        return ocorrencias;
    }

    public void setOcorrencias(ArrayList<Ocorrencia> ocorrencias) {
        this.ocorrencias = ocorrencias;
    }

    public void addOcorrencia(Ocorrencia ocorrencia) {
        this.ocorrencias.add(ocorrencia);
    }

    public void addPublicacao(Publicacao publicacao) {
        this.publicacoes.add(publicacao);
    }

    public ArrayList<Publicacao> getPublicacoes() {
        return publicacoes;
    }

    public void setPublicacoes(ArrayList<Publicacao> publicacoes) {
        this.publicacoes = publicacoes;
    }
}

File: Programação orientada a objetos/POO-241 - Trab 2 - Pedro_Quedevez/Docente.cpp

#include <iostream>
#include <string>
#include <vector>
#include <ctime>
#include "StringUtils.hpp"
#include "Docente.hpp"
#include "Tokenizer.hpp"
#include "DateUtils.hpp"
#include "Excecao.hpp"

using namespace cpp_util;
using namespace std;

Docente::Docente(string csvline)
{
    Tokenizer tokens(csvline, ';');
    vector<string> campos = tokens.remaining();

    for (int i = 0; i < campos.size(); i++)
    {
        campos[i] = trim(campos[i]);
    }

    try
    {

        if (codigosUtilizados.find(trim(campos[0])) != codigosUtilizados.end())
        {
            throw Excecao("Código repetido para docente: " + campos[0] + ".");
        }
        this->setCodigo(trim(campos[0]));
        this->setNome(trim(campos[1]));

        Tokenizer tokens(campos[2], '/');
        vector<string> dataNascimento = tokens.remaining();
        this->setAnoNascimento(stoi(dataNascimento[2]));

        Tokenizer tokens2(campos[3], '/');
        vector<string> dataIngresso = tokens2.remaining();
        this->setAnoIngresso(stoi(dataIngresso[2]));

        this->setDataNascimento(campos[2]);
        this->setDataIngresso(campos[3]);

        this->ocorrencias = vector<Ocorrencia>();
        this->publicacoes = vector<Publicacao>();

        codigosUtilizados.insert(trim(campos[0]));
    }

    catch (Excecao &e)
    {
        throw e;
    }
}

int Docente::idadeDocente(int anoAtual)
{
    int diferencaAnos = anoAtual - this->getAnoNascimento();

    return (diferencaAnos - 1);
}

int Docente::tempoIngresso(int anoAtual)
{

    int diferencaAnos = anoAtual - this->getAnoIngresso();

    return diferencaAnos;
}

void Docente::addOcorrencia(Ocorrencia o)
{
    this->ocorrencias.push_back(o);
}

void Docente::addPublicacao(Publicacao p)
{
    this->publicacoes.push_back(p);
}
string Docente::getCodigo()
{
    return this->codigo;
}
string Docente::getNome()
{
    return this->nome;
}
void Docente::setCodigo(string s)
{
    this->codigo = s;
}
void Docente::setNome(string s)
{
    this->nome = s;
}
time_t Docente::getDataNascimento()
{
    return this->dataNascimento;
}
time_t Docente::getDataIngresso()
{
    return this->dataIngresso;
}
void Docente::setDataNascimento(time_t t)
{
    this->dataNascimento = t;
}
void Docente::setDataIngresso(time_t t)
{
    this->dataIngresso = t;
}

void Docente::setAnoNascimento(int ano)
{
    this->anoNascimento = ano;
}

int Docente::getAnoNascimento()
{
    return this->anoNascimento;
}

void Docente::setAnoIngresso(int ano)
{
    this->anoIngresso = ano;
}

int Docente::getAnoIngresso()
{
    return this->anoIngresso;
}

void Docente::setDataNascimento(string s)
{
    this->dataNascimento = cpp_util::parseDate(trim(s), cpp_util::DATE_FORMAT_PT_BR_SHORT);
}
void Docente::setDataIngresso(string s)
{
    this->dataIngresso = cpp_util::parseDate(trim(s), cpp_util::DATE_FORMAT_PT_BR_SHORT);
}

vector<Ocorrencia> Docente::getOcorrencias()
{
    return this->ocorrencias;
}

vector<Publicacao> Docente::getPublicacoes()
{
    return this->publicacoes;
}

ostream &operator<<(ostream &c, Docente &d)
{
    c << d.getCodigo() << " " << d.getNome() << " " << cpp_util::formatDate(d.getDataNascimento(), cpp_util::DATE_FORMAT_PT_BR_SHORT) << " " << cpp_util::formatDate(d.getDataIngresso(), cpp_util::DATE_FORMAT_PT_BR_SHORT);
    return c;
}
